src/toast/spt3g/spt3g_import.py

```diff
@@ -0,0 +1,707 @@
+# Copyright (c) 2015-2021 by the parties listed in the AUTHORS file.
+# All rights reserved.  Use of this source code is governed by
+# a BSD-style license that can be found in the LICENSE file.
+
+import os
+import io
+import re
+import datetime
+
+import numpy as np
+
+import h5py
+
+from astropy import units as u
+
+from ..utils import Environment, Logger, import_from_name
+
+from ..timing import function_timer
+
+from ..instrument import Focalplane, GroundSite, SpaceSite
+
+from ..weather import SimWeather
+
+from ..observation import Observation
+
+from ..intervals import IntervalList
+
+from .spt3g_utils import (
+    available,
+    from_g3_array_type,
+    from_g3_scalar_type,
+    from_g3_unit,
+    from_g3_time,
+    from_g3_quats,
+    decompress_timestream,
+)
+
+if available:
+    from spt3g import core as c3g
+
+
+def check_obs_range(obs, offset, nsamp):
+    """Check that the requested sample range is valid."""
+    log = Logger.get()
+    if offset < 0 or offset >= obs.n_local_samples:
+        msg = f"observation {obs.name}:  sample offset {offset} "
+        msg += f"is outside the local range (0 - {obs.n_local_samples})"
+        log.error(msg)
+        raise RuntimeError(msg)
+
+    if offset + nsamp < 0 or offset + nsamp > obs.n_local_samples:
+        msg = f"observation {obs.name}:  end of samples "
+        msg += f" {offset + nsamp} "
+        msg += f"is outside the local range (0 - {obs.n_local_samples})"
+        log.error(msg)
+        raise RuntimeError(msg)
+
+
+def import_shared(obs, name, offset, g3obj):
+    """Copy a single G3Object into an observation shared field.
+
+    Args:
+        obs (Observation):  The parent observation.
+        name (str):  The name of the shared object.
+        offset (int):  The local data starting offset.
+        g3obj (G3Object):  The G3Object to copy.
+
+    Returns:
+        None
+
+    """
+    if name not in obs.shared:
+        raise KeyError(f"Shared object '{name}' does not exist in observation")
```

Comment:
Why can't we just create the field?

Reply:
This is a helper function that is copying a single frame's worth of one G3 object (i.e. a timestream, vector, etc) into a larger (pre-created) shared data object.  When importing, the schema of the first frame is used to determine all the shared / detdata objects that need to be created.  After these are pre-created in the observation, we go through each frame and each object and copy that small slice of data into the larger buffer.